GenderPayGap.UnitTests/GenderPayGap.WebUI.Tests/CodeQualityTests/DoNotUseRedirectWithReturnUrls.cs

```diff
@@ -0,0 +1,64 @@
+﻿using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using GenderPayGap.WebUI.Tests.TestHelpers;
+using NUnit.Framework;
+
+namespace GenderPayGap.WebUI.Tests.CodeQualityTests
+{
+    [TestFixture]
+    public class DoNotUseRedirectWithReturnUrls
+    {
+
+        [Test]
+        public void DoNotAllowRedirectWithReturnUrlInCode()
+        {
+            var excludedFilesAndFolders = new List<string>
+            {
+                @"GenderPayGap.UnitTests\GenderPayGap.WebUI.Tests\CodeQualityTests\DoNotUseRedirectWithReturnUrls.cs", // This file!
+                @"GenderPayGap.WebUI\bin", // Output folder
+                @"GenderPayGap.WebUI\obj" // Output folder
+            };
+
+            // Arrange
+            string rootCodeFolder = CodeQualityTestHelpers.GetRootCodeFolder();
+
+            // Pre-Act Assert (to check we're running the test on the right folder)
+            Assert.That(File.Exists($"{rootCodeFolder}\\GenderPayGap.sln"), $"We expect to find a file [GenderPayGap.sln] in the root folder [{rootCodeFolder}]");
+
+            // More Arrange
+            string searchPattern = "*.cs*" /* We want to find .cs and .cshtml files */;
+            FileInfo[] files = new DirectoryInfo(rootCodeFolder).GetFiles(searchPattern, SearchOption.AllDirectories);
+
+            // Pre-Act Assert (to check again that we're running the test on the right folder)
+            Assert.Greater(files.Length, 1000, "We expect there to be >1000 .cs/.cshtml files");
+
+            // Act
+            var failedFiles = new List<string>();
+            foreach (FileInfo fileInfo in files)
+            {
+                string filePath = fileInfo.FullName;
+                string filePathSuffix = filePath.Replace(rootCodeFolder, "");
+
+                if (CodeQualityTestHelpers.FileIsExcluded(filePathSuffix, excludedFilesAndFolders))
+                {
+                    continue;
+                }
+
+                string fileText = File.ReadAllText(filePath);
+                // There are a couple of actions on controllers Redirect() that wrap a RedirectToAction,
+                // we want to exclude these from the check as we are only concerned about redirecting with a url parameter
+                if (fileText.Contains(" Redirect(") && !fileText.Contains(" Redirect()"))
```

Comment:
I'm not mega keen on this check, I'm concerned e.g. you'll get edge cases with the wrong sort of whitespace at the start (a regex check would be better).

More importantly, I'm not sure this checks the right thing - it enforces future developers use `new RedirectResult` rather than `Redirect`, but what we really want to do is enforce that they have used `LocalRedirect` if possible (and if it's a legitimate external redirect, they can mark it as such and this test won't fail).

Ideally I think this check should flag if
* there's a `Redirect` or `new RedirectResult` which isn't a `LocalRedirect`
* _and_ they haven't stuck a comment on the previous line saying `// disable:DoNotAllowRedirectWithReturnUrlInCode` or something

And then the failure text should inform the user they should either switch to a `LocalRedirect` or stick the comment in there if they understand the risks. WDYT?

Reply:
Yeah I'm not super happy about it either and just followed the pattern of the Html.Raw test. Will look into how to do it a different way.

Do you have any particular thoughts on the use of Redirect vs new RedirectResult? If I change the unit test, I can change those so that they are more consistent.

Reply:
Good question. Sounds like they're equivalent so security wise I don't mind - but if it makes the test easier to write you could enforce we consistently use one of the two? Up to you!

Reply:
Updated to use a regex check to pick up instances of 'Redirect(url)' or 'new RedirectResult(url)' and included the check for a comment on the previous line to ignore an instance.